### PR TITLE
BanditParameter implementation

### PR DESCRIPTION
**Pull request by ekalosak**
*Issue #, if available:*
NA

*Description of changes:*
Parameter that represents a subset of the full Cartesian product of its subparameters.
Needs tests and an associated acquisition optimizer change (optimizing on the latent full Cartesian product of subparameters and rounding to nearest element of the valid subset is no bueno) before merge, but the Parameter implementation is ready for review.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by ekalosak** [comment]
Ok @apaleyes, I think I've addressed your comments.

Furthermore, using the RandomSearch acquisition optimizer assuages my concerns about using the Gradient acquisition optimizer with highly restricted design spaces, and there're unit tests for the BanditParameter now. Cursory in-house benchmarking and integration testing has supplied some anecdotal stress- and performance-test data - no catastrophic failures 👍

File: emukit/core/bandit_parameter.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
+# Copyright 2020 Opsani, Inc. or its affiliates. All Rights Reserved.
```
Comment (apaleyes) [inline emukit/core/bandit_parameter.py:1]: I honestly have no idea what the rules are around copyright notices. We'll need someone from Amazon to check on this one (I am no longer working there to check myself).

Comment (MashaNaslidnyk) [reply emukit/core/bandit_parameter.py:1]: Looking into it.

File: emukit/core/bandit_parameter.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        """
+        if x.ndim != 2:
+            raise ValueError("Expected 2d array, got " + str(x.ndim))
+
+        if x.shape[1] != self.dimension:
+            raise ValueError("Expected {} column array, got {}".format(self.dimension, x.shape[1]))
+
+        x_rounded = []
+        for row in x:
+            dists = np.sqrt(np.sum((self.domain - row)**2))
```
Comment (MashaNaslidnyk) [inline emukit/core/bandit_parameter.py:114]: is numerical stability a concern here? (can the value under sqrt be small and negative)

File: tests/emukit/core/test_parameters.py [hunk 3]
```diff
@@ -1,8 +1,34 @@
...
+from emukit.core import (
+    BanditParameter,
+    ContinuousParameter,
+    DiscreteParameter,
+    InformationSourceParameter,
+)
+
+
+def test_bandit_parameter():
+    # domain = np.array([('a',1),('a',2),('b',2)], dtype=[('x0','S1'),('x1','i4')])
```
Comment (MashaNaslidnyk) [inline tests/emukit/core/test_parameters.py:13]: remove?

File: emukit/core/bandit_parameter.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        """
+        Rounds each row in x to represent a valid value for this bandit variable. Note that this
+        valid value may be 'far' from the suggested value.
+
+        :param x: A 2d array NxD to be rounded (D is len(self.parameters))
+        :returns: An array NxD where each row represents a value from the domain
+                  that is closest to the corresponding row in x
+        """
+        if x.ndim != 2:
+            raise ValueError("Expected 2d array, got " + str(x.ndim))
```
Comment (MashaNaslidnyk) [inline emukit/core/bandit_parameter.py:107]: consistency: "Expected 2d array, got {}".format(x.ndim)

File: emukit/core/bandit_parameter.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+                parameter = DiscreteParameter(name = parameter_name, domain = domain_unq)
+            else:  # make categorical
+                encoding = OneHotEncoding(domain_unq)
+                parameter = CategoricalParameter(name = parameter_name, encoding = encoding)
+                raise NotImplementedError("Categorical sub-parameters not yet fully supported")
+                # NOTE Categorical sub-parameters not yet implemented because inputs are
+                # homogeneously typed np.ndarrays rather than structured arrays. In the future,
+                # using structured arrays for all inputs may be more appropriate.
+            parameters.append(parameter)
+        return(parameters)
```
Comment (MashaNaslidnyk) [inline emukit/core/bandit_parameter.py:58]: consistency: return parameters

File: emukit/core/bandit_parameter.py (hunk 2 above)
Comment (ekalosak) [reply emukit/core/bandit_parameter.py:114]: Unless the `self.domain` and `row` values are complex, those are sums of squares of real numbers. So no the summands will not be negative -> the sum will be non-negative.
Numerical stability could be a concern here, and I'm pretty sure there's `numpy.linalg.norm` which is a potentially better option.

File: emukit/core/bandit_parameter.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    A multivariate parameter consisting of a restricted domain of the full Cartesian product of its
+    constituent sub-parameters
+    """
+    def __init__(self, name: str, domain: np.ndarray, sub_parameter_names: Optional[List[str]]=None):
+        """
+        :param name: Name of parameter
+        :param domain: List of tuples representing valid values
+        :param parameters: List of parameters, must correspond to domain if provided, otherwise will
+        be reflected from the domain
+        """
```
[the remark this answers is not in the document]
Comment (ekalosak) [reply emukit/core/bandit_parameter.py:26]: Yeah you're right, fixing...

File: emukit/core/bandit_parameter.py (hunk 2 above)
Comment (MashaNaslidnyk) [reply emukit/core/bandit_parameter.py:114]: ah, of course, didn't notice that, sorry. 
I recall scipy.linalg.norm is more stable but I think it's fine as is.

**Comment by ekalosak** [comment]
Hey @MashaNaslidnyk @apaleyes - I've addressed the PR comments (thank you!) and have rebased onto master. Please give this another 👁️ at earliest convenience.

File: emukit/core/bandit_parameter.py (hunk 1 above)
Comment (MashaNaslidnyk) [reply emukit/core/bandit_parameter.py:1]: Checked, this is fine. Thanks for flagging Andrei!

Comment (apaleyes) [reply emukit/core/bandit_parameter.py:1]: I am pleasantly surprised. Thanks for checking @MashaNaslidnyk

**Comment by apaleyes** [comment]
@MashaNaslidnyk thanks for taking care of this! Just a small thing. Going forward please consider using "Squash and merge" for mergin pull requests. The main point is to keep the commit history clean, cause while working on PRs authors are free to do whatever they wish, and it isn't necessarily worth preserving.

**Pull request merged**